### ISSUE TITLE
铁子，代码写的有待提高

### ISSUE THREAD

**Issue by DrugD** [qinmingyuan-hub/molecular_graph#1]
#     def forward(self, smiles):
#         # import pdb; pdb.set_trace()
#         atom_graph,bond_graph = create_atom_bond_graph(smiles, self.args, self.args.force_field)
#         # atom_graph = create_atom_graph(smiles, self.args)
#         # bond_graph = create_bond_graph(smiles, self.args)

#         atoms_feature, atom_index = atom_graph.get_atom_feature() # Features, index records
#         bonds_feature, bond_index = bond_graph.get_bond_feature()

#         fpn_out = self.fpn(smiles)
#         self.device = fpn_out.device
#         smiles_egnn_cls=[]

#         atoms_feature = atoms_feature.to(self.device)
#         bonds_feature = bonds_feature.to(self.device)
#         fpn_out = fpn_out.to(self.device)

#         for i, one in enumerate(smiles):  # Connection matrix
#             mol = Chem.MolFromSmiles(one)
#             atom2bond_list = []
#             atom_edges = []
#             for x, bond in enumerate(mol.GetBonds()): # Atom graph
#                 atom1 = bond.GetBeginAtomIdx()
#                 atom2 = bond.GetEndAtomIdx()
#                 atom2bond_list.append((x, (atom1, atom2))) # Record the edge index and the connected atoms
#                 atom_edges.extend([[atom1, atom2], [atom2, atom1]]) # Add molecular graph
#             bond_edges = []
#             for m in range(len(atom2bond_list)): # Bond graph
#                 for n in range(m + 1, len(atom2bond_list)): # Traverse pairwise, if there are common atoms, generate an edge
#                     if set(atom2bond_list[m][1]) & set(atom2bond_list[n][1]):
#                         # bond_edges.append((m, n)) #
#                         bond_edges.extend([[m, n], [n, m]]) # Bidirectional
#             atom_edges = torch.tensor(atom_edges, dtype=torch.long).t().contiguous()
#             bond_edges = torch.tensor(bond_edges, dtype=torch.long).t().contiguous()  # New bond graph

#             atom_edges = atom_edges.to(self.device)
#             bond_edges = bond_edges.to(self.device)

#             atom_start, atom_size = atom_index[i]
#             bond_start, bond_size = bond_index[i]

#             one_atom_feature = atoms_feature[atom_start:atom_start + atom_size]  # Features of the i-th molecule

#             one_atom_feature_3d = one_atom_feature[:, -3:]
#             one_atom_feature_scalar = one_atom_feature[:, :-3]

#             # Feature convolution first
#             one_atom_feature_scalar = self.GCN_atom(one_atom_feature_scalar, atom_edges)

#             one_atom_feature_scalar, one_atom_feature_3d, global_atom_tokens = self.atom_3d_net(one_atom_feature_scalar, one_atom_feature_3d, atom_edges)

#             # Apply 1D average pooling
#             atom_concatenated_cls = torch.sum(global_atom_tokens[0] * self.weights_atom, dim=0, keepdim=True)

#             # atom_concatenated_cls = torch.flatten(global_atom_tokens, start_dim=1)
#             # bond_concatenated_cls = torch.flatten(global_bond_tokens, start_dim=1)
#             if bond_size!=0:
#                 one_bond_feature = bonds_feature[bond_start:bond_start + bond_size]  # Features of the i-th molecule
#                 one_bond_feature_3d = one_bond_feature[:, -3:]
#                 one_bond_feature_scalar = one_bond_feature[:, :-3]  # Bond convolution and then merge into atom
#                 one_bond_feature_scalar = self.GCN_bond(one_bond_feature_scalar, bond_edges)
#                 transformer_bond_out = self.bond_3d_net(one_bond_feature_scalar) # Change to pure transformer encoder

#             else:
#                 transformer_bond_out=torch.zeros(1,self.linear_dim).to(self.device)

#             smiles_egnn_cls.append(torch.cat([atom_concatenated_cls,transformer_bond_out], axis=1))

#         # import pdb; pdb.set_trace()
#         smiles_egnn_cls = torch.stack(smiles_egnn_cls, dim=1)[0]

#         smiles_egnn_cls = self.ffn_atom_bond(smiles_egnn_cls)
#         padded_batch_data = torch.cat([smiles_egnn_cls,fpn_out], axis=1)

#         output = self.ffn(padded_batch_data)  # Define ffn according to different scales
#         if self.is_classif and not self.training:
#             output = self.sigmoid(output)
#         return padded_batch_data, output
这forward看得我大吃一惊，能不能改一版不是一个smiles一个smiles处理的正常的那种合理用上GPU的代码（-：
